src/t_zset.c

```diff
@@ -2184,6 +2200,177 @@ inline static void zunionInterAggregate(double *target, double val, int aggregat
     }
 }
 
+static int zsetDictGetMaxElementLength(dict *d) {
+    dictIterator *di;
+    dictEntry *de;
+    size_t maxelelen = 0;
+
+    di = dictGetIterator(d);
+
+    while((de = dictNext(di)) != NULL) {
+        sds ele = dictGetKey(de);
+        if (sdslen(ele) > maxelelen) maxelelen = sdslen(ele);
+    }
+
+    dictReleaseIterator(di);
+
+    return maxelelen;
+}
+
+static void zdiffAlgorithm1(zsetopsrc *src, long setnum, zset *dstzset, size_t *maxelelen) {
+    /* DIFF Algorithm 1:
+     *
+     * We perform the diff by iterating all the elements of the first set,
+     * and only adding it to the target set if the element does not exist
+     * into all the other sets.
+     *
+     * This way we perform at max N*M operations, where N is the size of
+     * the first set, and M the number of sets.
+     *
+     * There is also a O(K*log(K)) cost for adding the resulting elements
+     * to the target set, where K is the final size of the target set.
+     *
+     * The final complexity of this algorithm is O(N*M + K*log(K)). */
+    int j;
+    zsetopval zval;
+    zskiplistNode *znode;
+    sds tmp;
+
+    /* With algorithm 1 it is better to order the sets to subtract
+     * by decreasing size, so that we are more likely to find
+     * duplicated elements ASAP. */
+    qsort(src+1,setnum-1,sizeof(zsetopsrc),zuiCompareByRevCardinality);
+
+    memset(&zval, 0, sizeof(zval));
+    zuiInitIterator(&src[0]);
+    while (zuiNext(&src[0],&zval)) {
```

Comment:
going over the list backwards will make zslInsert more efficient, see:
https://github.com/redis/redis/blob/d8fbd3a8bafc9b430da6c3c13bdd04af4d75ad0c/src/rdb.c#L855-L862

alternatively, maybe we can improve zslInsert to be optimized for this case too?

Reply:
Ok, I just analyzed the code, and it's a bit more complicated than I thought. The referenced code already knows that it's a skiplist. In the case of zdiff, we have to cover all cases since the `src` values can have any encoding. I've checked that `zuiInitIterator` and `zuiNext` are only used in zinter/zunion/zdiff, so I thought about changing then to `zuiInitBackIterator` and `zuiPrev`. I'll have to study the data structure a bit more to understand how to do that, but shouldn't be too hard. I'll try to finish this tomorrow.